[v2] Add configuration to disable host prefix injection

kdaily opened the pull request:
*Issue #, if available:*

*Description of changes:*

This change deviates from the upstream `botocore` to add the ability to disable jost prefix injection. It changes the default for `inject_host_prefix` in the Config object to be `None` (from `True`) so that the value can be determined using the configuration precedence. The fallback if the value remains `True` if it is set to `None` to preserve the CLI default behavior.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

hssyoo reviewed (commented):
Looks good, just a couple small suggestions. I'm also curious if there are existing unit/functional tests where you can actually set the environment variable or create a config file to test that they resolve correctly E2E.

ashovlin reviewed (commented):
Could we wait to merge until we figure out the plan for botocore? Want to understand where V1 vs. V2 will land.

kdaily commented on awscli/botocore/args.py:
```diff
@@ -212,7 +212,6 @@ def compute_client_args(self, service_model, client_config,
                 proxies_config=client_config.proxies_config,
                 retries=client_config.retries,
                 client_cert=client_config.client_cert,
-                inject_host_prefix=client_config.inject_host_prefix,
```
Sure, not opposed to this. I went for self-containing all of the logic in the `compute` method.

kdaily commented on .changes/next-release/feature-configuration-42674.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "type": "feature",
+  "category": "configuration",
+  "description": "Configure if the host prefix is prepended to the request URL in the shared configuration file or via an environment variable"
```
Still need to do this - need to regenerate changelog after refactoring.

kdaily commented:
> Could we wait to merge until we figure out the plan for botocore? Want to understand where V1 vs. V2 will land.

We should still wait to merge once we confirm that this is going to be fine for `botocore`.

kdaily commented:
> Looks good, just a couple small suggestions. I'm also curious if there are existing unit/functional tests where you can actually set the environment variable or create a config file to test that they resolve correctly E2E.

I think we can - this should have been done for configured endpoint URLs. I'll investigate the same still.

kdaily commented:
Re: functional testing when actually setting in config file or as an env variable: here's the examples we did for the configured endpoint URL. As I recall we did this specifically because this change also introduced complexity with the `services` section of the config file. Other params are not tested as extensively beyond what I already added here.

- In `botocore`: https://github.com/aws/aws-cli/blob/v2/tests/functional/botocore/configured_endpoint_urls/test_configured_endpoint_url.py
- In `aws-cli`: https://github.com/kdaily/aws-cli/blob/v2/tests/functional/configured_endpoint_urls/test_configured_endpoint_url.py

hssyoo commented on awscli/botocore/utils.py:
```diff
@@ -3980,3 +3980,11 @@ def is_s3express_bucket(bucket):
 }
+
+
+class DefaultTrue:
+    def __eq__(self, other):
+        return other == True
```
1 minor concern with this sentinel is that it doesn't work for identity operators (eg `DefaultTrue() is True` evaluates to `False`).

This is also a workaround to compensate for resolving configured values and I'm not sure we should be legitimizing it and potentially encouraging future contributors to use it.

What do you think of @ashovlin's suggestion of storing a public and private `inject_host_prefix` attribute in the `Config` class? We would have to compare against a private attribute in `args.py`, but it seems marginally better than replacing the default with this sentinel value.

kdaily commented on awscli/botocore/args.py:
```diff
@@ -256,6 +257,30 @@ def compute_client_args(self, service_model, client_config,
         }
 
+    def _compute_inject_host_prefix(self, config_kwargs):
+        # If the value isn't available, that is because the user didn't pass in a Config
+        # object. UNSET is used to differentiate between other currently allowed values,
+        #  like None.
+        inject_host_prefix = config_kwargs.get('inject_host_prefix', 'UNSET')
```
Usually one would do:

```
        inject_host_prefix = config_kwargs.get('inject_host_prefix', None)
```

But the current behavior allows passing `None`. Having a default return of `None` in the `get` would not be able to distinguish if it was supplied by the user or not.

hssyoo replied:
Ah I see what you're saying. But wouldn't creating `Config(inject_host_prefix=None)` be equivalent to saying "I don't want to set a client config-level value for this. So either use a value in the config store if it exists, otherwise use the default."

ashovlin replied:
We could change that over here if necessary though, right? Only pop if it's actually `False`?
https://github.com/aws/aws-cli/blob/74486f39a23ca96c9302f97802f179ff457f6b26/awscli/botocore/client.py#L771-L772

kdaily commented:
Marking as draft pending review and approval of PR in `botocore`:

https://github.com/boto/botocore/pull/3405
